File: nxflix-agents-ts/src/models/video.ts
import { z } from 'zod';

/**
 * Character/Avatar style selection for videos.
 */
export const CharacterStyle = z.enum([
  'anime_female',
  'anime_male',
  'realistic_female',
  'realistic_male',
  'chibi',
  'mascot',
  'none', // Subtitles only, no character
]);
export type CharacterStyle = z.infer<typeof CharacterStyle>;

/**
 * Video visual style/background.
 */
export const VideoStyle = z.enum([
  'classroom', // Traditional learning setting
  'cafe', // Casual conversation setting
  'nature', // Outdoor/scenic backgrounds
  'abstract', // Minimalist/gradient backgrounds
  'manga', // Comic panel style
]);
export type VideoStyle = z.infer<typeof VideoStyle>;

/**
 * Video project status.
 */
export const VideoStatus = z.enum([
  'draft',
  'generating',
  'ready',
  'failed',
]);
export type VideoStatus = z.infer<typeof VideoStatus>;

/**
 * Furigana reading annotation for a word.
 */
export const FuriganaAnnotation = z.object({
  word: z.string(), // Original word (kanji)
  reading: z.string(), // Hiragana reading
  startIndex: z.number().int(), // Position in text
});
export type FuriganaAnnotation = z.infer<typeof FuriganaAnnotation>;

/**
 * Subtitle with furigana support.
 */
export const VideoSubtitle = z.object({
  id: z.string(),
  startTime: z.number(), // Seconds from start
  endTime: z.number(),
  text: z.string(), // Japanese text
  reading: z.string().optional(), // Full reading for TTS
  furigana: z.array(FuriganaAnnotation).default([]),
  translation: z.string().optional(), // English translation
});
export type VideoSubtitle = z.infer<typeof VideoSubtitle>;

/**
 * Video script with timing and metadata.
 */
export const VideoScript = z.object({
  id: z.string(),
  title: z.string(),
  description: z.string().optional(),
  subtitles: z.array(VideoSubtitle),
  totalDurationSeconds: z.number(),
  targetVocabulary: z.array(z.string()).default([]),
  grammarPoints: z.array(z.string()).default([]),
});
export type VideoScript = z.infer<typeof VideoScript>;

/**
 * Video project with all assets.
 */
export const VideoProject = z.object({
  id: z.string(),
  userId: z.string(),
  prompt: z.string(), // Original user prompt
  script: VideoScript,
  characterStyle: CharacterStyle,
  videoStyle: VideoStyle,
  voice: z.string(), // TTS voice ID
  status: VideoStatus,
  audioUrl: z.string().optional(),
  audioBase64: z.string().optional(),
  videoUrl: z.string().optional(),
  thumbnailUrl: z.string().optional(),
  createdAt: z.string(),
  updatedAt: z.string(),
  errorMessage: z.string().optional(),
  progress: z.number().min(0).max(100).default(0),
});
export type VideoProject = z.infer<typeof VideoProject>;

/**
 * Request to create a new video project.
 */
export const VideoCreateRequest = z.object({
  prompt: z.string().min(1).max(500),
  userId: z.string().min(1),
  characterStyle: CharacterStyle.default('anime_female'),
  videoStyle: VideoStyle.default('classroom'),
  voice: z.string().optional(),
  maxDurationSeconds: z.number().int().min(15).max(60).default(60),
});
export type VideoCreateRequest = z.infer<typeof VideoCreateRequest>;

/**
 * Request to generate just a script.
 */
export const ScriptGenerateRequest = z.object({
  prompt: z.string().min(1).max(500),
  maxDurationSeconds: z.number().int().min(15).max(60).default(60),
});
export type ScriptGenerateRequest = z.infer<typeof ScriptGenerateRequest>;

/**
 * Available TTS voice info.
 */
export const TTSVoice = z.object({
  id: z.string(),
  name: z.string(),
  gender: z.string(),
  description: z.string(),
  provider: z.string(),
});
export type TTSVoice = z.infer<typeof TTSVoice>;

/**
 * Available style options response.
 */
export const VideoStylesResponse = z.object({
  characters: z.array(
    z.object({
      id: CharacterStyle,
      name: z.string(),
      description: z.string(),
    })
  ),
  backgrounds: z.array(
    z.object({
      id: VideoStyle,
      name: z.string(),
      description: z.string(),
    })
  ),
});
export type VideoStylesResponse = z.infer<typeof VideoStylesResponse>;

File: nxflix-agents-python/src/nxflix_agents/models/video.py
"""Video model for JLPT N1 educational content creation."""

from enum import Enum
from typing import Literal
from pydantic import BaseModel, Field


class CharacterStyle(str, Enum):
    """Character/Avatar style selection for videos."""

    ANIME_FEMALE = "anime_female"
    ANIME_MALE = "anime_male"
    REALISTIC_FEMALE = "realistic_female"
    REALISTIC_MALE = "realistic_male"
    CHIBI = "chibi"
    MASCOT = "mascot"
    NONE = "none"  # Subtitles only, no character


class VideoStyle(str, Enum):
    """Video visual style/background."""

    CLASSROOM = "classroom"  # Traditional learning setting
    CAFE = "cafe"  # Casual conversation setting
    NATURE = "nature"  # Outdoor/scenic backgrounds
    ABSTRACT = "abstract"  # Minimalist/gradient backgrounds
    MANGA = "manga"  # Comic panel style


class VideoStatus(str, Enum):
    """Video project status."""

    DRAFT = "draft"
    GENERATING = "generating"
    READY = "ready"
    FAILED = "failed"


class FuriganaAnnotation(BaseModel):
    """Furigana reading annotation for a word."""

    word: str  # Original word (kanji)
    reading: str  # Hiragana reading
    start_index: int  # Position in text


class VideoSubtitle(BaseModel):
    """Subtitle with furigana support."""

    id: str
    start_time: float  # Seconds from start
    end_time: float
    text: str  # Japanese text
    reading: str | None = None  # Full reading for TTS
    furigana: list[FuriganaAnnotation] = []
    translation: str | None = None  # English translation


class VideoScript(BaseModel):
    """Video script with timing and metadata."""

    id: str
    title: str
    description: str | None = None
    subtitles: list[VideoSubtitle]
    total_duration_seconds: float
    target_vocabulary: list[str] = []
    grammar_points: list[str] = []


class VideoProject(BaseModel):
    """Video project with all assets."""

    id: str
    user_id: str
    prompt: str  # Original user prompt
    script: VideoScript
    character_style: CharacterStyle
    video_style: VideoStyle
    voice: str  # TTS voice ID
    status: VideoStatus
    audio_url: str | None = None
    audio_base64: str | None = None
    video_url: str | None = None
    thumbnail_url: str | None = None
    created_at: str
    updated_at: str
    error_message: str | None = None
    progress: int = Field(default=0, ge=0, le=100)


class VideoCreateRequest(BaseModel):
    """Request to create a new video project."""

    prompt: str = Field(..., min_length=1, max_length=500)
    user_id: str = Field(..., min_length=1)
    character_style: CharacterStyle = CharacterStyle.ANIME_FEMALE
    video_style: VideoStyle = VideoStyle.CLASSROOM
    voice: str | None = None
    max_duration_seconds: int = Field(default=60, ge=15, le=60)


class ScriptGenerateRequest(BaseModel):
    """Request to generate just a script."""

    prompt: str = Field(..., min_length=1, max_length=500)
    max_duration_seconds: int = Field(default=60, ge=15, le=60)


class TTSVoice(BaseModel):
    """Available TTS voice info."""

    id: str
    name: str
    gender: str
    description: str
    provider: str


class StyleOption(BaseModel):
    """A style option for characters or backgrounds."""

    id: str
    name: str
    description: str


class VideoStylesResponse(BaseModel):
    """Available style options response."""

    characters: list[StyleOption]
    backgrounds: list[StyleOption]
